File: string-search/strategies/BoyerMooreSearch.js
import SearchStrategy from "./SearchStrategy.js";

export default class BoyerMooreSearch extends SearchStrategy {
  buildBadCharTable(pattern) {
    const table = {};

    for (let i = 0; i < pattern.length; i++) {
      table[pattern[i]] = i;
    }

    return table;
  }

  search(text, pattern, stepMode = false) {
    const steps = [];
    const matches = [];
    let comparisons = 0;

    if (!pattern.length || pattern.length > text.length) {
      return {
        algorithm: "Boyer-Moore",
        matches,
        comparisons,
        textLength: text.length,
        patternLength: pattern.length,
        complexity: "O(n / m) melhor caso",
        steps,
        auxiliary: { badChar: {} }
      };
    }

    const badChar = this.buildBadCharTable(pattern);
    let shift = 0;

    if (stepMode) {
      steps.push("Construindo tabela de bad character...");
    }

    while (shift <= text.length - pattern.length) {
      let j = pattern.length - 1;

      if (stepMode) {
        steps.push(`Iniciando comparação com shift = ${shift}`);
      }

      while (j >= 0) {
        comparisons++;

        if (stepMode) {
          steps.push(`Comparando text[${shift + j}] = "${text[shift + j]}" com pattern[${j}] = "${pattern[j]}"`);
        }

        if (pattern[j] !== text[shift + j]) {
          break;
        }

        j--;
      }

      if (j < 0) {
        matches.push(shift);

        if (stepMode) {
          steps.push(`Padrão encontrado na posição ${shift}`);
        }

        shift += shift + pattern.length < text.length
          ? pattern.length - (badChar[text[shift + pattern.length]] ?? -1)
          : 1;
      } else {
        const badCharIndex = badChar[text[shift + j]];
        const jump = Math.max(1, j - (badCharIndex ?? -1));

        if (stepMode) {
          steps.push(`Falha. Caractere ruim: "${text[shift + j]}". Salto de ${jump}`);
        }

        shift += jump;
      }
    }

    return {
      algorithm: "Boyer-Moore",
      matches,
      comparisons,
      textLength: text.length,
      patternLength: pattern.length,
      complexity: "O(n / m) melhor caso",
      steps,
      auxiliary: { badChar }
    };
  }
}

File: AlgoritmosDeBusca_String/app/strategies/boyer_moore_search.py
import time
from typing import Dict

from app.domain.search_result import SearchResult
from app.domain.search_strategy import SearchStrategy
from app.strategies._step_helper import add_step


class BoyerMooreSearch(SearchStrategy):
    @property
    def name(self) -> str:
        return "Boyer-Moore"

    @property
    def theoretical_complexity(self) -> str:
        return "O(n / m) melhor caso; O(n * m) pior caso"

    def _build_bad_character_table(self, pattern: str) -> Dict[str, int]:
        return {char: index for index, char in enumerate(pattern)}

    def search(
        self,
        text: str,
        pattern: str,
        include_steps: bool = False,
        max_steps: int = 300,
    ) -> SearchResult:
        start = time.perf_counter_ns()
        occurrences = []
        comparisons = 0
        steps = []

        n = len(text)
        m = len(pattern)

        if m == 0 or n == 0 or m > n:
            elapsed_ms = (time.perf_counter_ns() - start) / 1_000_000
            return SearchResult(
                algorithm=self.name,
                occurrences=[],
                comparisons=0,
                execution_time_ms=elapsed_ms,
                text_size=n,
                pattern_size=m,
                theoretical_complexity=self.theoretical_complexity,
                auxiliary_data={"bad_character_table": {}},
                steps=steps,
            )

        bad_character_table = self._build_bad_character_table(pattern)
        add_step(steps, f"Tabela Bad Character criada: {bad_character_table}.", include_steps, max_steps)

        shift = 0
        while shift <= n - m:
            j = m - 1
            add_step(steps, f"Alinhando padrão na posição {shift}.", include_steps, max_steps)

            while j >= 0:
                comparisons += 1
                add_step(
                    steps,
                    f"Comparando texto[{shift + j}]='{text[shift + j]}' com padrão[{j}]='{pattern[j]}'.",
                    include_steps,
                    max_steps,
                )

                if pattern[j] != text[shift + j]:
                    break
                j -= 1

            if j < 0:
                occurrences.append(shift)
                add_step(steps, f"Ocorrência encontrada na posição {shift}.", include_steps, max_steps)

                if shift + m < n:
                    next_char = text[shift + m]
                    shift_amount = m - bad_character_table.get(next_char, -1)
                else:
                    shift_amount = 1

                add_step(steps, f"Deslocando {shift_amount} posição(ões).", include_steps, max_steps)
                shift += shift_amount
            else:
                mismatched_char = text[shift + j]
                last_occurrence = bad_character_table.get(mismatched_char, -1)
                shift_amount = max(1, j - last_occurrence)
                add_step(
                    steps,
                    f"Falha com caractere '{mismatched_char}'. Deslocando {shift_amount} posição(ões).",
                    include_steps,
                    max_steps,
                )
                shift += shift_amount

        elapsed_ms = (time.perf_counter_ns() - start) / 1_000_000
        return SearchResult(
            algorithm=self.name,
            occurrences=occurrences,
            comparisons=comparisons,
            execution_time_ms=elapsed_ms,
            text_size=n,
            pattern_size=m,
            theoretical_complexity=self.theoretical_complexity,
            auxiliary_data={"bad_character_table": bad_character_table},
            steps=steps,
        )
